File: orbslam.py
import cv2
import numpy as np

class ORBSLAM():
    """
    Wrapper for the ORB feature detector and matcher
    """
    def __init__(self):
        self.orb = cv2.ORB_create()
        self.matcher = cv2.BFMatcher(cv2.NORM_HAMMING, crossCheck=True)
        self.prev = None

    def extract(self, frame):
        # find the keypoints and descriptors with ORB
        corners = cv2.goodFeaturesToTrack(frame, 0, 0.01, 10)
        kps = [cv2.KeyPoint(x=f[0][0], y=f[0][1], size=20) for f in corners]
        _, des = self.orb.compute(frame, kps)

        return kps, des

    def match(self, des1, des2):
        """Match descriptors from two images"""
        print(des1)
        matches = self.matcher.match(des1, des2)
        matches = sorted(matches, key = lambda x:x.distance)
        print(f"Found {len(matches)} matches")
        return matches

    def compute(self, prev_frame, frame):
        """Compute keypoints, descriptors and matches across frames"""
        # extract features
        kps, des = self.extract(frame)

        if prev_frame is None:
            # save the previous frame
            self.prev = {'kps': kps, 'des': des} 
            return kps, des, None

        # match features
        matches = self.match(des, self.prev['des'])

        # save the previous frame
        self.prev = {'kps': kps, 'des': des}
        return kps, des, matches

File: c++_implementation/src/orbslam.cpp
/**
 * @file orbslam.cpp
 * @author George O'Connor
 * @brief Class to to do useful orb slam things
 * @version 0.1
 * @date 2022-10-07
 *
 * @copyright Copyright (c) 2022
 *
 */

#include "orbslam.hpp"


ORBSLAM::ORBSLAM() {
    _orb = cv::ORB::create();
    _matcher = cv::DescriptorMatcher::create("BruteForce-Hamming");
};

ORBSLAM::~ORBSLAM() {
    _orb.release();
    _matcher.release();
};


std::tuple<std::vector<cv::KeyPoint>, cv::Mat>  ORBSLAM::extract(cv::Mat frame) {
    /**
     * @brief Extracts features from a frame
     *
     * @param frame
     * @return std::tuple<std::vector<cv::KeyPoint>, cv::Mat>
     */
    std::vector<cv::KeyPoint> keypoints;
    std::vector<cv::Point2f> corners;
    cv::goodFeaturesToTrack(frame, corners, 1000, 0.01, 10);
    cv::KeyPoint::convert(corners, keypoints, 10, 1, 0, -1);
    cv::Mat descriptors;
    _orb->compute(frame, keypoints, descriptors);
    return std::make_tuple(keypoints, descriptors);

}

std::vector<cv::DMatch> ORBSLAM::match(cv::Mat prev_descriptors, cv::Mat descriptors) {
    /**
     * @brief Matches features between two frames
     *
     * @param prev_descriptors
     * @param descriptors
     * @return std::vector<cv::DMatch>
     */
    std::vector<cv::DMatch> matches;
    _matcher->match(prev_descriptors, descriptors, matches);
    // sort matches by distance
    std::sort(matches.begin(), matches.end());
    return matches;
}

std::tuple<std::vector<cv::KeyPoint>, cv::Mat, std::vector<cv::DMatch>> ORBSLAM::compute(cv::Mat prev_frame, cv::Mat frame) {
    /**
     * @brief Extract and match feautres between the current frame and the previous frame
     * If there is no previous frame, then just extract features
     *
     * @param prev_frame
     * @param frame
     * @return std::tuple<std::vector<cv::KeyPoint>, cv:Mat, cv::DMatch>
     */
    std::vector<cv::KeyPoint> keypoints;
    cv::Mat descriptors;
    std::vector<cv::DMatch> matches;

    std::tie(keypoints, descriptors) = extract(frame);
    if (prev_frame.empty()) {
        _prev_descriptors = descriptors;
        _prev_keypoints = keypoints;
        return std::make_tuple(keypoints, descriptors, matches);
    }

    matches = match(_prev_descriptors, descriptors);
    _prev_descriptors = descriptors;
    _prev_keypoints = keypoints;
    return std::make_tuple(keypoints, descriptors, matches);
}

